design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,481 @@
+using K8s' [Downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/) mechanism.
+
+A new field will be added to the `NetworkInfo` struct:
+
+```go
+type NetworkInfo struct {
+Interfaces []Interface `json:"interfaces,omitempty"`
+PrimaryPodInterfaceName string `json:primaryPodInterfaceName,omitempty`
+}
+```
+
+A new mechanism should be added between virt-handler and virt-launcher to share this information, because:
+
+1. The primary pod interface name is also required by virt-handler
+2. The virt-handler does not have access to pod objects
+
+virt-handler will poll the virt-launcher's file-system for the `/etc/podinfo` file.
+virt-handler will not be able to start the network setup until it successfully read and parsed the content of the file above.
+
+#### Cons
+
+- All virt-launcher pods will have to use the downward API mechanism instead of just the subset using it today
+- Complicates the network setup and overall VM startup sequence
+
+## Specifying the primary pod interface name
+
+### Adding a cluster-wide preferred pod NIC name
+
+A new field will be added in KubeVirt's cluster-wide network configuration, which will specify the preferred pod
+interface name that should be bound to the VM interface destined for pod network:
+
+```go
+type NetworkConfiguration struct {
+...
+PreferredPodPrimaryInterfaceName string `json:"preferredPodPrimaryInterfaceName,omitempty"`
+...
+}
+```
+
+#### API Example
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  name: kubevirt
+  namespace: kubevirt
+spec:
+  configuration:
+    network:
+      preferredPodPrimaryInterfaceName: special-nic
+```
+
+virt-handler and virt-launcher will be adjusted to prefer this value when trying to bind the primary network interface.
+They will both fallback to `eth0` in case they cannot find the preferred interface name.
+
+#### Pros
+
+- Does not require Multus to work
+
+#### Cons
+
+- The cluster admin or an [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) needs to know in advance which interface name to use
+- Error prone
+- Has cluster-wide granularity
+- Requires a change to the communication protocol between virt-handler and virt-launcher to pass the cluster-wide
+  setting to virt-launcher
+
+### Specifying the primary pod NIC name in advance in K8s or on the node
+
+When creating a VM with secondary networks, KubeVirt specifies the pod interfaces’ names for all secondary interfaces.
+It is currently not possible to specify the primary interface’s name when templating the virt-launcher pod, since there
+is no K8s API that enables it.
+
+The primary pod interface name is set by the container runtime (for example cri-o), before invoking the default CNI.
+
+#### No Go
+
+- There is no K8s API to name the pod's primary NIC
+- There is no Multus API or config to name the pod's primary NIC
+
+## Network Binding Plugins
```

Comment:
it is not clear if this is another option or it extends the chosen option for network bindings.

If there are several options, please:
- Add the chosen one at the beginning.
- Add a ToC for the options/alternatives.